torch/csrc/jit/codegen/cuda/interface.cpp

```diff
@@ -34,11 +39,129 @@ void fuseGraph(std::shared_ptr<Graph>& graph) {
   getFuserInterface()->fn_fuse_graph(graph);
 }
 
+// [ Note -- type guard logic in CudaFusionGuard ]
+//
+// Type guard confirms that runtime `tensor` complies with profiled
+// `guard_tensor_type`.
+//
+// At this moment, we only do single profiling run, so `guard_tensor_type` has
+// static shape / stride / scalarType. *This might be a little confusing as our
+// implementation is actually more relaxed.
+//
+// Things that we check:
+//   a. identical rank & scalar type
+//   b. stride check:
+//        b.1. identical stride order
+//        b.2. identical contiguity
+//             note that contiguity here is used for tensor collapsing. So extra
+//             attention should be paid to contiguity across size-1 dimensions.
+//   c. size check:
+//        making sure that broadcast semantics are identical. So we want to make
+//        sure a given dimension either are both size-1 for `tensor` &
+//        `guard_tensor_type`, or are both non-size-1.
+//
+bool complyWith(
+    const at::Tensor& tensor,
+    const c10::TensorTypePtr& guard_tensor_type) {
+  // guard broadcast semantics, contiguity & stride order;
```

Comment:
should we add a `FUSER_PERF_SCOPE` here?

Reply:
I had that earlier but removed it, because I think that will unfortunately break CPU build.

interface.cpp is with cpu only build (because of op registration), while instrumentation is only included with cuda BUILD.
But there is not reason we can't pull `instrument.x` & `utils.x` to cpu dependency. I'll try that.